panel/pane/web_component.py

```diff
@@ -0,0 +1,424 @@
+        for prop, pname in self.properties_to_watch.items():
+            if pname not in msg:
+                continue
+            value = msg.pop(pname)
+            if msg.get('propertiesLastChange') is None:
+                msg['propertiesLastChange'] = {}
+            msg['propertiesLastChange'][prop] = value
+        for pname in self.events_to_watch.values():
+            if pname in msg:
+                del msg[pname]
+        html_change = any(pname in msg for pname in self.parameters_to_watch)
+        if html_change:
+            msg = {k: v for k, v in msg.items() if k not in self.parameters_to_watch}
+            params = {parameter: getattr(self, parameter) for parameter in self.parameters_to_watch}
+            html = self._get_html_from_parameters_to_watch(**params)
+            self.html = html
+
+        return msg
+
+    def _process_property_change(self, msg):
+        msg = super(WebComponent, self)._process_property_change(msg)
+        if 'html' in msg:
+            attr_dict = self.parse_html_to_dict(msg['html'])
+            for attribute, parameter in self.attributes_to_watch.items():
+                if not parameter:
+                    continue
+                ptype = type(self.param[parameter])
+                if ptype not in PARAMETER_TYPE:
+                    raise TypeError("Parameter of type '%s' is not supported." % ptype)
+                if attribute in attr_dict:
+                    attr_value = attr_dict[attribute]
+                    value = PARAMETER_TYPE[ptype](attr_value)
+                else:
+                    value = self.param[parameter].default
+                msg[attribute] = value
+        if 'attributes_last_change' in msg:
+            for attr, new_value in msg['attributes_last_change'].items():
+                panme = self.attributes_to_watch.get(attr)
+                if pname is None:
+                    continue
+                pobj = self.param[pname]
+                ptype = type(pobj)
+                msg[pname] = PARAMETER_TYPE[ptype](new_value)
+        if 'events_count_last_change' in msg:
+            for prop, new_value in msg['events_count_last_change'].items():
+                pname = self.events_to_watch.get(prop)
+                if pname is None:
+                    continue
+                pobj = self.param[pname]
+                ptype = type(pobj)
+                msg[pname] = PARAMETER_TYPE[ptype](new_value)
+        if 'properties_last_change' in msg:            
+            for prop, new_value in msg['properties_last_change'].items():
+                pname = self.properties_to_watch.get(prop)
+                if pname is None:
+                    continue
+                pobj = self.param[pname]
+                ptype = type(pobj)
+                if isinstance(new_value, str) and ptype in PARAMETER_TYPE:
+                    new_value = PARAMETER_TYPE[ptype](new_value)
+                msg[pname] = new_value            
```

Comment:
@MarcSkovMadsen I've refactored all the different methods you had introduced to make of the messaging protocols that Panel itself provides. This makes it much clearer what the differences in serialization and deserialization approaches are. Can you explain to me why properties, events and attributes have subtly different deserialization behaviors? Can these be unified? As example the `properties_last_change` deserialization explicitly checks:

```
if isinstance(new_value, str)
```

Is this necessary? 'events' and 'attributes' don't have this check.

Reply:
There are differences. For example for a boolean value on an attribute.

If True its stated as `attribute` or `attribute=""`. If False the attribute is just removed.

Maybe some things can be cleaned up. Hopefully the many tests will catch changes. The problem is that I don't know how to setup Typescript tests. So there are none.

I just remember it was complicated. But the many examples I created helped find all the corner cases.